pkg/hsperfdata/hsperfdata.go

```diff
@@ -0,0 +1,144 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+
+package hsperfdata
+
+import (
+	"fmt"
+	"io/fs"
+	"os"
+	"path/filepath"
+	"strconv"
+	"strings"
+	"sync"
+
+	"github.com/go-kit/log"
+
+	"golang.org/x/sync/singleflight"
+
+	"github.com/parca-dev/parca-agent/pkg/perf"
+)
+
+const hsperfdata = "/tmp/hsperfdata_*"
+
+type Cache struct {
+	pids   map[int]struct{}
+	fs     fs.FS
+	logger log.Logger
+	mu     sync.Mutex
+	nsPID  map[int]int
+	group  singleflight.Group
+}
+
+type realfs struct{}
+
+func (f *realfs) Open(name string) (fs.File, error) {
+	return os.Open(name)
+}
+
+func NewCache(logger log.Logger) *Cache {
+	return &Cache{
+		pids:   make(map[int]struct{}),
+		fs:     &realfs{},
+		logger: logger,
+		nsPID:  map[int]int{},
+	}
+}
+
+func (c *Cache) Get(pid int) bool {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+
+	_, ok := c.pids[pid]
+	return ok
+}
+
+// IsJavaProcess returns true if the hsperfdata file exists for a given pid.
+// It first searches in all hsperfdata user directories for the processes
+// running on host and then searches in /proc/{pid}/root/tmp for processes
+// running in containers. Note that pids are assumed to be unique regardless
+// of username.
+func (c *Cache) IsJavaProcess(pid int) (bool, error) {
+	// Check if the pid is in the cache
+	if c.Get(pid) {
+		return true, nil
+	}
+
+	// Use singleflight to prevent concurrent requests for the same pid
+	ch := c.group.DoChan(strconv.Itoa(pid), func() (interface{}, error) {
+		// List all directories that match the pattern /tmp/hsperfdata_*
+		dirs, err := filepath.Glob(hsperfdata)
+		if err != nil {
+			return false, fmt.Errorf("failed to list directories: %w", err)
+		}
+
+		// Loop over all directories and search for the hsperfdata file for the given pid
+		for _, dir := range dirs {
+			hsperfdataPath := filepath.Join(dir, strconv.Itoa(pid))
+			if _, err := c.fs.Open(hsperfdataPath); err == nil {
+				c.pids[pid] = struct{}{}
+				return true, nil
+			}
+		}
+
+		// Search for the pid via nspids.
+		nsPid, found := c.nsPID[pid]
+		if !found {
+			nsPids, err := perf.FindNSPIDs(c.fs, pid)
+			if err != nil {
+				if os.IsNotExist(err) {
+					return false, fmt.Errorf("%w when reading status", perf.ErrProcNotFound)
+				}
+				return false, err
+			}
+
+			c.nsPID[pid] = nsPids[len(nsPids)-1]
+			nsPid = c.nsPID[pid]
+		}
+
+		// TODO(vthakkar): Check for the process mount point.
+		perfdataFiles := fmt.Sprintf("/proc/%d/root/tmp/", pid)
+
+		files, err := os.ReadDir(perfdataFiles)
+		if err != nil {
+			return false, fmt.Errorf("error reading %s: %w", perfdataFiles, err)
+		}
+
+		for _, f := range files {
+			if f.IsDir() {
+				if name := f.Name(); strings.HasPrefix(name, "hsperfdata") {
+					if _, err := c.fs.Open(filepath.Join(perfdataFiles, name, strconv.Itoa(nsPid))); err == nil {
+						c.pids[pid] = struct{}{}
+						return true, nil
+					}
+				}
+			}
+		}
+
+		c.pids[pid] = struct{}{}
```

Comment:
This case is for when we can't find the hsperf file, right? In this case why do we need to add this pid to the "is pid JAVA" cache?

Reply:
Oops, good catch. That was some personal debugging leftover :see_no_evil: Will wait for Kemal's review and push the fixes after that. Thanks for the review!